text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,325 @@
+
+The actions it adds may be generic (from level 2), translatable into any backend and so
+the approval workflow is reusable, or they may be backend-specific and only apply to
+one specific backend.
+
+```
+┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+                                                   App
+│┌────────────────────────────────────────────────────┐          ┌───────────────────────────────────┐  │
+ │                       Deploy                       │          │              Approve              │
+││                                                    │          │                                   │  │
+ │ ┌────────────────────┐    ┌────────────────────┐   │          │                                   │
+││ │       Stack1       │    │       Stack2       │   │          │                                   │  │
+ │ │                    │    │                    │   │          │                                   │
+││ │  ┌────┐    ┌────┐  │    │  ┌────┐    ┌────┐  │   │─────────▶│                                   │  │
+ │ │  │ C  │───▶│ E  │  │───▶│  │ C  │───▶│ E  │  │   │          │                                   │
+││ │  └────┘    └────┘  │    │  └────┘    └────┘  │   │          │                                   │  │
+ │ │                    │    │                    │   │          │                                   │
+││ └────────────────────┘    └────────────────────┘   │          │                                   │  │
+ │                                                    │          │                                   │
+│└────────────────────────────────────────────────────┘          └───────────────────────────────────┘  │
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+```
+
+### How will asset publishing work?
+
+Similar to how approval workflows work, the asset publishing strategy can be configured
+with a callback which can manipulate the workflow as desired. This allows for easy switching
+between:
+
+* Prepublish all assets in individual CodeBuild projects
+* Prepublish all assets in one CodeBuild project
+* Publish assets just before each app deployment
+* Publish initial assets as usual but afterwards wait for ECR replication to finish
+
+### What are the generic primitives?
+
+In the middle layer, the following types of actions/states exist:
+
+* ShellScript
+  - May have concepts like shell commands, Docker image, IAM permissions it should
+    be possible to assume
+  - Renders to CodeBuild for CodePipeline, a Step in GitHub Actions
+* Manual Approval
+  - They GitHub Actions renderer may reject this type of action, and that is okay.
+* Create ChangeSet
+* Execute ChangeSet
+  - With or without capturing outputs (backend may reject capturing outputs if it cannot implement that)
+
+In addition to these, a specific implementation may add backend-specific actions.
```

Comment:
This is what I was expecting as the pattern to support backend specific things in the top-layer.